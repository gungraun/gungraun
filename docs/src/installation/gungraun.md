# Gungraun

Gungraun is divided into the library `gungraun` and the benchmark runner
`gungraun-runner`.

## Installation of the library

To start with Gungraun, add the following to your `Cargo.toml` file:

```toml
[dev-dependencies]
gungraun = "0.16.1"
```

or run

```bash
cargo add --dev gungraun@0.16.1
```

## Installation of the benchmark runner

To be able to run the benchmarks you'll also need the `gungraun-runner`
binary installed somewhere in your `$PATH`. Otherwise, there is no need to
interact with `gungraun-runner` as it is just an implementation detail.

### From Source

```shell
cargo install --version 0.16.1 gungraun-runner
```

There's also the possibility to install the binary somewhere else and point the
`GUNGRAUN_RUNNER` environment variable to the absolute path of the
`gungraun-runner` binary like so:

```shell
cargo install --version 0.16.1 --root /tmp gungraun-runner
GUNGRAUN_RUNNER=/tmp/bin/gungraun-runner cargo bench --bench my-bench
```

### Binstall

The `gungraun-runner` binary is
[pre-built](https://github.com/gungraun/gungraun/releases/tag/v0.16.1)
for most platforms supported by valgrind and easily installable with
[binstall](https://github.com/cargo-bins/cargo-binstall)

```shell
cargo binstall gungraun-runner@0.16.1
```

## Updating

When updating the `gungraun` library, you'll also need to update
`gungraun-runner` and vice-versa or else the benchmark runner will exit
with an error.

### In the Github CI

Since the `gungraun-runner` version must match the `gungraun` library
version it's best to automate this step in the CI. A job step in the github
actions CI could look like this

```yaml
- name: Install gungraun-runner
  run: |
    version=$(cargo metadata --format-version=1 |\
      jq '.packages[] | select(.name == "gungraun").version' |\
      tr -d '"'
    )
    cargo install gungraun-runner --version $version
```

Or, speed up the overall installation time with `binstall` using the
[taiki-e/install-action](https://github.com/taiki-e/install-action)

```yaml
- uses: taiki-e/install-action@cargo-binstall
- name: Install gungraun-runner
  run: |
    version=$(cargo metadata --format-version=1 |\
      jq '.packages[] | select(.name == "gungraun").version' |\
      tr -d '"'
    )
    cargo binstall --no-confirm gungraun-runner --version $version
```
