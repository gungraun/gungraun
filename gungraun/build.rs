//! The build script

// spell-checker: ignore rustified iquote iquotevalgrind

#[cfg(feature = "client_requests_defs")]
mod imp {
    use std::borrow::Cow;
    use std::fmt::Display;
    use std::io::{BufRead, BufReader, Cursor};
    use std::path::PathBuf;

    use bindgen::{builder, Bindings};
    use rustc_version::{version, Version};
    use strum::{EnumIter, IntoEnumIterator};

    #[derive(Debug)]
    struct Target {
        arch: String,
        env: String,
        os: String,
        triple: String,
        vendor: String,
    }

    #[derive(EnumIter, Debug, PartialEq, Eq)]
    enum Support {
        Arm,
        Aarch64,
        X86,
        X86_64,
        Riscv64,
        Native,
        No,
    }

    impl Display for Support {
        fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
            let support = format!("{self:?}").to_lowercase();
            f.write_str(&support)
        }
    }

    impl Target {
        fn from_env() -> Self {
            Self {
                arch: std::env::var("CARGO_CFG_TARGET_ARCH").unwrap(),
                env: std::env::var("CARGO_CFG_TARGET_ENV").unwrap(),
                os: std::env::var("CARGO_CFG_TARGET_OS").unwrap(),
                vendor: std::env::var("CARGO_CFG_TARGET_VENDOR").unwrap(),
                triple: std::env::var("TARGET").unwrap(),
            }
        }
    }

    fn print_client_requests_support(value: &Support) {
        println!("cargo:rustc-cfg=client_requests_support=\"{value}\"");
    }

    fn include_dirs(target: &Target) -> impl Iterator<Item = String> {
        [
            Cow::Owned(format!(
                "IAI_CALLGRIND_{}_VALGRIND_INCLUDE",
                target.triple.replace('-', "_").to_ascii_uppercase()
            )),
            Cow::Borrowed("IAI_CALLGRIND_VALGRIND_INCLUDE"),
        ]
        .into_iter()
        .filter_map(|env| std::env::var(env.as_ref()).ok())
    }

    fn build_native(target: &Target) {
        let mut builder = cc::Build::new();

        for env in include_dirs(target) {
            builder.include(env);
        }

        if target.os == "freebsd" {
            builder.include("/usr/local/include");
        }

        if let Ok(env) = std::env::var("GUNGRAUN_CROSS_TARGET") {
            let path = PathBuf::from("/valgrind/target/valgrind")
                .join(env)
                .join("include");
            builder.include(path);
        }

        builder.include("valgrind/include");

        builder
            .debug(true)
            .file("valgrind/native.c")
            .compile("native");
    }

    fn build_bindings(target: &Target) -> Bindings {
        let mut builder = builder();

        for env in include_dirs(target) {
            builder = builder.clang_arg(format!("-iquote{env}"));
        }

        if let Ok(env) = std::env::var("GUNGRAUN_CROSS_TARGET") {
            let path = PathBuf::from("/valgrind/target/valgrind")
                .join(env)
                .join("include");
            builder = builder.clang_arg(format!("-iquote{}", path.display()));
        }

        if target.os == "freebsd" {
            builder = builder.clang_arg("-iquote/usr/local/include");
        }

        let bindings = builder
            .clang_arg("-iquote/usr/include")
            .clang_arg("-iquotevalgrind/include")
            .header("valgrind/wrapper.h")
            .allowlist_var("IC_IS_PLATFORM_SUPPORTED_BY_VALGRIND")
            .allowlist_var("IC_VALGRIND_MAJOR")
            .allowlist_var("IC_VALGRIND_MINOR")
            .allowlist_type("IC_.*ClientRequest")
            .rustified_enum("IC_.*ClientRequest")
            .layout_tests(false)
            .parse_callbacks(Box::new(bindgen::CargoCallbacks::new()))
            .generate()
            .expect("Generating binding should succeed");

        let out_dir = std::env::var("OUT_DIR").map(PathBuf::from).unwrap();
        let path = out_dir.join("bindings.rs");
        bindings.write_to_file(path).unwrap();
        bindings
    }

    // Return the rust version if running rustc was successful
    fn get_rust_version() -> Option<Version> {
        version().ok()
    }

    pub fn main() {
        println!("cargo:rerun-if-changed=valgrind/wrapper.h");
        println!("cargo:rerun-if-changed=valgrind/native.c");
        println!("cargo:rerun-if-env-changed=IAI_CALLGRIND_VALGRIND_INCLUDE");
        println!("cargo:rerun-if-env-changed=GUNGRAUN_CROSS_TARGET");
        println!("cargo:rerun-if-env-changed=TARGET");

        // rustc-check-cfg is introduced in rust with version 1.80 and avoids the compiler warnings
        // in version >= 1.80.0. Printing it when compiling with versions < 1.80 triggers a warning,
        // too. To get the best of both worlds we check against the currently active rust version.
        if let Some(rust_version) = get_rust_version() {
            if rust_version.major >= 1 && rust_version.minor >= 80 {
                let values = Support::iter()
                    .map(|s| format!("\"{s}\""))
                    .collect::<Vec<String>>()
                    .join(",");
                println!("cargo:rustc-check-cfg=cfg(client_requests_support,values({values}))");
            }
        }

        let target = Target::from_env();

        // When building the docs on docs.rs we can take a shortcut
        if std::env::var("DOCS_RS").is_ok() {
            print_client_requests_support(&Support::X86_64);
            build_bindings(&target);
            build_native(&target);
            return;
        }

        let bindings = build_bindings(&target);

        // These guards mirror the checks in the `valgrind.h` header file
        let support = if target.arch == "x86_64"
            && (target.os == "linux"
                || target.os == "freebsd"
                || (target.vendor == "apple" && target.os == "darwin")
                || (target.os == "windows" && target.env == "gnu")
                || ((target.vendor == "sun") || target.vendor == "pc") && target.os == "solaris")
        {
            Some(Support::X86_64)
        } else if target.arch == "x86"
            && (target.os == "linux"
                || target.os == "freebsd"
                || (target.vendor == "apple" && target.os == "darwin")
                || (target.os == "windows" && target.env == "gnu")
                || ((target.vendor == "sun") || target.vendor == "pc") && target.os == "solaris")
        {
            Some(Support::X86)
        } else if target.arch == "arm" && target.os == "linux" && target.env == "gnu" {
            Some(Support::Arm)
        } else if target.arch == "aarch64"
            && (target.os == "freebsd" || (target.os == "linux" && target.env == "gnu"))
        {
            Some(Support::Aarch64)
        } else if target.arch == "riscv64gc" && target.os == "linux" {
            Some(Support::Riscv64)
        } else {
            let re = regex::Regex::new(
                r"IC_IS_PLATFORM_SUPPORTED_BY_VALGRIND.*?=\s*(?<value>true|false)",
            )
            .expect("Regex should compile");
            let reader = BufReader::new(Cursor::new(bindings.to_string()));
            let mut support = None;
            for line in reader.lines().map(Result::unwrap) {
                if let Some(caps) = re.captures(&line) {
                    let value = caps.name("value").unwrap().as_str();
                    if value == "false" {
                        support = Some(Support::No);
                    } else if value == "true" {
                        support = Some(Support::Native);
                    } else {
                        // do nothing
                    }
                    break;
                }
            }
            support
        };

        if let Some(support) = support {
            print_client_requests_support(&support);
            if support != Support::No {
                build_native(&target);
            }
        } else {
            eprintln!("{bindings}");
            panic!("Unable to set cfg value for client_requests_support");
        }
    }
}

#[cfg(not(feature = "client_requests_defs"))]
mod imp {
    pub fn main() {}
}

fn main() {
    imp::main();
}
