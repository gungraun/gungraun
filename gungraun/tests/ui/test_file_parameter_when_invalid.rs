use gungraun::{binary_benchmark, library_benchmark};

mod test_empty_file {
    use super::*;

    #[library_benchmark]
    #[benches::my_id(file = "gungraun/tests/fixtures/empty.fix")]
    fn bench_library(value: String) -> u64 {
        value.parse::<u64>().unwrap()
    }

    #[binary_benchmark]
    #[benches::my_id(file = "gungraun/tests/fixtures/empty.fix")]
    fn bench_binary(value: String) -> u64 {
        value.parse::<u64>().unwrap()
    }
}

mod test_file_does_not_exist {
    use super::*;

    #[library_benchmark]
    #[benches::my_id(file = "gungraun/tests/fixtures/does_not_exist")]
    fn bench_library(value: String) -> u64 {
        value.parse::<u64>().unwrap()
    }

    #[binary_benchmark]
    #[benches::my_id(file = "gungraun/tests/fixtures/does_not_exist")]
    fn bench_binary(value: String) -> u64 {
        value.parse::<u64>().unwrap()
    }
}

mod test_invalid_utf8 {
    use super::*;

    #[library_benchmark]
    #[benches::my_id(file = "gungraun/tests/fixtures/invalid-utf8.fix")]
    fn bench_library(value: String) -> u64 {
        value.parse::<u64>().unwrap()
    }

    #[binary_benchmark]
    #[benches::my_id(file = "gungraun/tests/fixtures/invalid-utf8.fix")]
    fn bench_binary(value: String) -> u64 {
        value.parse::<u64>().unwrap()
    }
}

mod test_args_and_file_parameter {
    use super::*;

    #[library_benchmark]
    #[benches::my_id(file = "gungraun/tests/fixtures/numbers.fix", args = [("valid_arg".to_owned()), "another".to_owned()])]
    fn bench_library(value: String) -> u64 {
        value.parse::<u64>().unwrap()
    }

    #[binary_benchmark]
    #[benches::my_id(file = "gungraun/tests/fixtures/numbers.fix", args = [("valid_arg".to_owned()), "another".to_owned()])]
    fn bench_binary(value: String) -> u64 {
        value.parse::<u64>().unwrap()
    }
}

mod test_wrong_benchmark_argument_type {
    use super::*;

    #[library_benchmark]
    #[benches::my_id(file = "gungraun/tests/fixtures/numbers.fix")]
    fn bench_library(value: u64) -> String {
        value.to_string()
    }

    #[binary_benchmark]
    #[benches::my_id(file = "gungraun/tests/fixtures/numbers.fix")]
    fn bench_binary(value: u64) -> String {
        value.to_string()
    }
}

mod test_wrong_parameter_type_1 {
    use super::*;

    #[library_benchmark]
    #[benches::my_id(file = ("gungraun/tests/fixtures/numbers.fix", String))]
    fn bench_library(value: String) -> u64 {
        value.parse::<u64>().unwrap()
    }

    #[binary_benchmark]
    #[benches::my_id(file = ("gungraun/tests/fixtures/numbers.fix", String))]
    fn bench_binary(value: String) -> u64 {
        value.parse::<u64>().unwrap()
    }
}

mod test_wrong_parameter_type_2 {
    use super::*;

    #[library_benchmark]
    #[benches::my_id(file = String::from("gungraun/tests/fixtures/numbers.fix"))]
    fn bench_library(value: String) -> u64 {
        value.parse::<u64>().unwrap()
    }

    #[binary_benchmark]
    #[benches::my_id(file = String::from("gungraun/tests/fixtures/numbers.fix"))]
    fn bench_binary(value: String) -> u64 {
        value.parse::<u64>().unwrap()
    }
}

mod test_wrong_amount_of_benchmark_function_parameters {
    use super::*;

    #[library_benchmark]
    #[benches::my_id(file = "gungraun/tests/fixtures/numbers.fix")]
    fn bench_library(value: String, _other: String) -> u64 {
        value.parse::<u64>().unwrap()
    }

    #[binary_benchmark]
    #[benches::my_id(file = "gungraun/tests/fixtures/numbers.fix")]
    fn bench_binary(value: String, _other: String) -> u64 {
        value.parse::<u64>().unwrap()
    }
}

mod test_wrong_benchmark_function_parameters_when_setup {
    use super::*;

    fn my_setup(line: String) -> u64 {
        line.parse().unwrap()
    }

    #[library_benchmark]
    #[benches::my_id(file = "gungraun/tests/fixtures/numbers.fix", setup = my_setup)]
    fn bench_library(value: String) -> u64 {
        value.parse::<u64>().unwrap()
    }

    #[binary_benchmark]
    #[benches::my_id(file = "gungraun/tests/fixtures/numbers.fix", setup = { my_setup("some string".to_owned()); })]
    fn bench_binary(value: u64) -> gungraun::Command {
        gungraun::Command::new("nope")
    }
}

fn main() {}
