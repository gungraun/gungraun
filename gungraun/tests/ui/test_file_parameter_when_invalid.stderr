error: The provided file 'gungraun/tests/fixtures/empty.fix' was empty
 --> tests/ui/test_file_parameter_when_invalid.rs:7:29
  |
7 |     #[benches::my_id(file = "gungraun/tests/fixtures/empty.fix")]
  |                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: The provided file 'gungraun/tests/fixtures/empty.fix' was empty
  --> tests/ui/test_file_parameter_when_invalid.rs:13:29
   |
13 |     #[benches::my_id(file = "gungraun/tests/fixtures/empty.fix")]
   |                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: Error opening '$WORKSPACE/target/tests/trybuild/gungraun/gungraun/tests/fixtures/does_not_exist': No such file or directory (os error 2)
  --> tests/ui/test_file_parameter_when_invalid.rs:23:29
   |
23 |     #[benches::my_id(file = "gungraun/tests/fixtures/does_not_exist")]
   |                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: Error opening '$WORKSPACE/target/tests/trybuild/gungraun/gungraun/tests/fixtures/does_not_exist': No such file or directory (os error 2)
  --> tests/ui/test_file_parameter_when_invalid.rs:29:29
   |
29 |     #[benches::my_id(file = "gungraun/tests/fixtures/does_not_exist")]
   |                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: Error reading line 1 in file '$WORKSPACE/target/tests/trybuild/gungraun/gungraun/tests/fixtures/invalid-utf8.fix': stream did not contain valid UTF-8
  --> tests/ui/test_file_parameter_when_invalid.rs:39:29
   |
39 |     #[benches::my_id(file = "gungraun/tests/fixtures/invalid-utf8.fix")]
   |                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: Error reading line 1 in file '$WORKSPACE/target/tests/trybuild/gungraun/gungraun/tests/fixtures/invalid-utf8.fix': stream did not contain valid UTF-8
  --> tests/ui/test_file_parameter_when_invalid.rs:45:29
   |
45 |     #[benches::my_id(file = "gungraun/tests/fixtures/invalid-utf8.fix")]
   |                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: Only one parameter of `file`, `args` or `iter` can be present
  --> tests/ui/test_file_parameter_when_invalid.rs:55:16
   |
55 |     #[benches::my_id(file = "gungraun/tests/fixtures/numbers.fix", args = [("valid_arg".to_owned()), "another".to_owned()])]
   |                ^^^^^

error: Only one parameter of `file`, `args` or `iter` can be present
  --> tests/ui/test_file_parameter_when_invalid.rs:61:16
   |
61 |     #[benches::my_id(file = "gungraun/tests/fixtures/numbers.fix", args = [("valid_arg".to_owned()), "another".to_owned()])]
   |                ^^^^^

error: Invalid value for `file`

         = help: The `file` argument needs a literal string containing the path to an existing file at compile time
         = note: `file = "benches/some_fixture"`

  --> tests/ui/test_file_parameter_when_invalid.rs:87:29
   |
87 |     #[benches::my_id(file = ("gungraun/tests/fixtures/numbers.fix", String))]
   |                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: Invalid value for `file`

         = help: The `file` argument needs a literal string containing the path to an existing file at compile time
         = note: `file = "benches/some_fixture"`

  --> tests/ui/test_file_parameter_when_invalid.rs:93:29
   |
93 |     #[benches::my_id(file = ("gungraun/tests/fixtures/numbers.fix", String))]
   |                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: Invalid value for `file`

         = help: The `file` argument needs a literal string containing the path to an existing file at compile time
         = note: `file = "benches/some_fixture"`

   --> tests/ui/test_file_parameter_when_invalid.rs:103:29
    |
103 |     #[benches::my_id(file = String::from("gungraun/tests/fixtures/numbers.fix"))]
    |                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: Invalid value for `file`

         = help: The `file` argument needs a literal string containing the path to an existing file at compile time
         = note: `file = "benches/some_fixture"`

   --> tests/ui/test_file_parameter_when_invalid.rs:109:29
    |
109 |     #[benches::my_id(file = String::from("gungraun/tests/fixtures/numbers.fix"))]
    |                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: The benchmark function should take exactly one `String` argument if the file parameter is present

         = help: fn benchmark_function(line: String) ...

   --> tests/ui/test_file_parameter_when_invalid.rs:119:29
    |
119 |     #[benches::my_id(file = "gungraun/tests/fixtures/numbers.fix")]
    |                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: The benchmark function should take exactly one `String` argument if the file parameter is present

         = help: fn benchmark_function(line: String) ...

   --> tests/ui/test_file_parameter_when_invalid.rs:125:29
    |
125 |     #[benches::my_id(file = "gungraun/tests/fixtures/numbers.fix")]
    |                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0308]: mismatched types
  --> tests/ui/test_file_parameter_when_invalid.rs:71:29
   |
71 |     #[benches::my_id(file = "gungraun/tests/fixtures/numbers.fix")]
   |                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |                             |
   |                             expected `u64`, found `String`
   |                             arguments to this function are incorrect
   |
help: the return type of this call is `String` due to the type of the argument passed
  --> tests/ui/test_file_parameter_when_invalid.rs:71:29
   |
71 |     #[benches::my_id(file = "gungraun/tests/fixtures/numbers.fix")]
   |                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ this argument influences the return type of `std`
note: function defined here
  --> $RUST/core/src/hint.rs
   |
   | pub const fn black_box<T>(dummy: T) -> T {
   |              ^^^^^^^^^

error[E0308]: mismatched types
  --> tests/ui/test_file_parameter_when_invalid.rs:77:29
   |
77 |     #[benches::my_id(file = "gungraun/tests/fixtures/numbers.fix")]
   |                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `u64`, found `String`
78 |     fn bench_binary(value: u64) -> String {
   |        ------------ arguments to this function are incorrect
   |
note: function defined here
  --> tests/ui/test_file_parameter_when_invalid.rs:78:8
   |
78 |     fn bench_binary(value: u64) -> String {
   |        ^^^^^^^^^^^^ ----------

error[E0308]: mismatched types
  --> tests/ui/test_file_parameter_when_invalid.rs:76:5
   |
76 |     #[binary_benchmark]
   |     ^^^^^^^^^^^^^^^^^^^
   |     |
   |     expected `Command`, found `String`
   |     expected `gungraun::Command` because of return type
   |
   = note: this error originates in the attribute macro `binary_benchmark` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0308]: mismatched types
   --> tests/ui/test_file_parameter_when_invalid.rs:138:5
    |
138 |     #[library_benchmark]
    |     ^^^^^^^^^^^^^^^^^^^^- help: try using a conversion method: `.to_string()`
    |     |
    |     expected `String`, found `u64`
    |     arguments to this function are incorrect
    |
note: function defined here
   --> tests/ui/test_file_parameter_when_invalid.rs:138:5
    |
138 |     #[library_benchmark]
    |     ^^^^^^^^^^^^^^^^^^^^
139 |     #[benches::my_id(file = "gungraun/tests/fixtures/numbers.fix", setup = my_setup)]
140 |     fn bench_library(value: String) -> u64 {
    |                      -------------
    = note: this error originates in the attribute macro `library_benchmark` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0308]: mismatched types
   --> tests/ui/test_file_parameter_when_invalid.rs:145:29
    |
145 |     #[benches::my_id(file = "gungraun/tests/fixtures/numbers.fix", setup = { my_setup("some string".to_owned()); })]
    |                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `u64`, found `String`
146 |     fn bench_binary(value: u64) -> gungraun::Command {
    |        ------------ arguments to this function are incorrect
    |
note: function defined here
   --> tests/ui/test_file_parameter_when_invalid.rs:146:8
    |
146 |     fn bench_binary(value: u64) -> gungraun::Command {
    |        ^^^^^^^^^^^^ ----------
